controllers/scaler_controller.go

```diff
@@ -0,0 +1,309 @@
+	csvOcsOperator           = "ocs-operator"
+	csvOdfCsiAddonsOperator  = "odf-csi-addons-operator" //nolint:unused
+	csvOdfDependencies       = "odf-dependencies"        //nolint:unused
+	csvOdfOperator           = "odf-operator"            //nolint:unused
+	csvOdfPrometheusOperator = "odf-prometheus-operator" //nolint:unused
+	csvRecipe                = "recipe"                  //nolint:unused
+	csvRookCephOperator      = "rook-ceph-operator"
+
+	csvIbmStorageOdfOperator = "ibm-storage-odf-operator"
+)
+
+var (
+	crdToKindMapping = map[string]metav1.PartialObjectMetadata{
+		crdStorageCluster:     *kindStorageCluster,
+		crdCephCluster:        *kindCephCluster,
+		crdNoobaa:             *kindNoobaa,
+		crdFlashSystemCluster: *kindFlashSystemCluster,
+	}
+
+	kindToCsvsMapping = map[*metav1.PartialObjectMetadata][]string{
+		kindStorageCluster:     []string{csvOcsOperator},
+		kindCephCluster:        []string{csvRookCephOperator},
+		kindNoobaa:             []string{csvMcgOperator},
+		kindFlashSystemCluster: []string{csvIbmStorageOdfOperator},
+	}
+)
+
+type ScalerReconciler struct {
+	ctx               context.Context
+	logger            logr.Logger
+	Client            client.Client
+	Scheme            *runtime.Scheme
+	OperatorNamespace string
+	controller        controller.Controller
+	mgr               ctrl.Manager
+}
+
+//+kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch
+//+kubebuilder:rbac:groups=operators.coreos.com,resources=clusterserviceversions,verbs=get;list;update
+
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=noobaa.io,resources=noobaas,verbs=get;list;watch
+//+kubebuilder:rbac:groups=odf.ibm.com,resources=flashsystemclusters,verbs=get;list;watch
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *ScalerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.logger = log.FromContext(ctx)
+
+	r.logger.Info("starting reconcile")
+
+	err := r.addWatches(req)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.scaleOperators()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.logger.Info("ending reconcile")
+	return ctrl.Result{}, nil
+}
+
+func (r *ScalerReconciler) scaleOperators() error {
+
+	for kind, csvNames := range kindToCsvsMapping {
+
+		var replicas int32 = 1
+		objects := &metav1.PartialObjectMetadataList{}
+		objects.TypeMeta = kind.TypeMeta
+
+		err := r.Client.List(r.ctx, objects)
+		if err != nil {
+			if meta.IsNoMatchError(err) {
+				continue
+			}
+			r.logger.Error(err, "failed to list objects")
+			return err
+		}
+
+		if len(objects.Items) == 0 {
+			replicas = 0
+		}
+
+		for _, csvName := range csvNames {
```

Comment:
maybe it's better to run this in parallel or let the loop go over all CSVs trying to update CSV as a failure in one stops everything else now, why do want to put a break if all operators can be run in parallel in the cluster?

Reply:
done